### PR TITLE
Implement numpy-like function `torch.block`

### PR DESCRIPTION
**Pull request by HimariO**
Reference: #38349
Implement numpy-like function `torch.block` with similar approach to `torch.block_diag`. Not sure if there is a better way to handle the nested list of tensors than flatten it inside `functional.py `.

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1644,6 +1644,144 @@ Tensor& dstack_out(TensorList tensors, Tensor& result) {
...
+      tensors[0].options().dtype(output_scalar_type));
+
+  for (auto i : c10::irange(tensors.size())) {
+    auto slice_a = slices_a[i];
+    auto slice_b = slices_b[i];
+    Tensor tensor = tensors[i];
+    Tensor ten_slic = result;
+    for (auto j: c10::irange(result_dim))
+      ten_slic = ten_slic.slice(j, slice_a[j], slice_b[j]);
+    ten_slic.copy_(tensor);
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:1779]: This looks a bit questionable, in particular it doesn't seem like autograd is going to work with an implementation like this.

Comment (HimariO) [reply aten/src/ATen/native/TensorShape.cpp:1779]: You are right on that, after some testing, it does show that autograd's won't able to propagate gradient through **slices**. I think creating a block_backward to deal with the gradient will be the most reasonable solution for this?

**Review by mruberry — COMMENTED** [review]
Hey @HimariO! 

This is pretty cool but this PR needs tests! New operators in the `torch` namespace should create an OpInfo, following the note here:

https://github.com/pytorch/pytorch/blob/849c6a526e86e66e5112407a53c6b052fd5342cd/torch/testing/_internal/common_methods_invocations.py#L258

The OpInfo should use NumPy's `block()` operation as a reference operator. Additional testing will likely be needed. One common type of test is to basically have an extended set of sample inputs and verify the output of `torch.block()` is consistent with the output of `np.block()`.

**Comment by HimariO** [comment]
Hi @mruberry, I had taken a look into the OpInfos and found out some limitations of the framework prevent me from implementing the test in a "normal" way. Be more specifically, `SampleInput` doesn't support the Tensor contained in the nested list so the test is unable to create the input needed for `torch.block`. I currently have a workaround of creating an auxiliary function `torch.block_flat` that take flatten Tensor list and creates corresponding OpInfo. Do you have any advice on this issue?

**Pull request closed without merging**